frame/contracts/src/lib.rs

```diff
@@ -695,24 +695,37 @@ pub mod pallet {
 			salt: Vec<u8>,
 		) -> DispatchResultWithPostInfo {
 			Migration::<T>::ensure_migrated()?;
+			let origin = ensure_signed(origin)?;
 			let code_len = code.len() as u32;
+
+			let (module, deposit) = Self::try_upload_code(
+				origin.clone(),
+				code,
+				storage_deposit_limit.clone().map(Into::into),
+				Determinism::Enforced,
+				None,
+			)?;
+
 			let data_len = data.len() as u32;
 			let salt_len = salt.len() as u32;
 			let common = CommonInput {
-				origin: Origin::from_runtime_origin(origin)?,
+				origin: Origin::from_account_id(origin),
 				value,
 				data,
 				gas_limit,
-				storage_deposit_limit: storage_deposit_limit.map(Into::into),
+				storage_deposit_limit: storage_deposit_limit
+					.map(|limit| limit.into().saturating_sub(deposit)),
```

Comment:
alright, this might need some comments for clarity: we've taken the deposit for storing the code above, that's why we reduce the limit appropriately here.

Reply:
Yes, I realize that tests pass with or without this, so I should add an extra test for this

Reply:
added a test that would fail without this condition